src/include/storage/write_ahead_log/log_io.h

```diff
@@ -95,44 +95,54 @@ class BufferedLogWriter {
 
   /**
    * Write to the log file the given amount of bytes from the given location in memory, but buffer the write so the
-   * update is only written out when the BufferedLogWriter is persisted.
+   * update is only written out when the BufferedLogWriter is persisted. Note that this function writes to the buffer
+   * only until it is full. If buffer gets full, then call FlushBuffer() and call BufferWrite(..) again with the correct
+   * offset of the data, depending on the number of bytes that were already written.
    * @param data memory location of the bytes to write
    * @param size number of bytes to write
+   * @return number of bytes written. This function only writes until the buffer gets full, so this can be used as the
+   * offset when calling this function again after flushing.
    */
-  void BufferWrite(const void *data, uint32_t size) {
-    if (!CanBuffer(size)) FlushBuffer();
-    // If we still do not have buffer space after flush, the write is too large to be buffered. We should bypass the
-    // buffer and write directly to disk
+  uint32_t BufferWrite(const void *data, uint32_t size) {
```

Comment:
This method is part of the BufferedLogWriter though, which owns only a single buffer, so it can't do much once it fulls up its buffer. It's up to the caller to find a new buffer to put the rest of the data in. What you're describing is [`LogManager::WriteValue`](https://github.com/cmu-db/terrier/pull/370/files#diff-70736a64a01e1c14098ca447fcbbb6d5R163), which is the only place `BufferWrite` is called.